File: src/main/java/br/com/pnp/compiler/model/construct/statement/DoWhileStatement.java
package br.com.pnp.compiler.model.construct.statement;

import br.com.pnp.compiler.model.construct.Block;
import br.com.pnp.compiler.model.construct.type.Type;
import br.com.pnp.compiler.model.construct.type.primitives.PrimitiveType;
import br.com.pnp.compiler.model.expression.Expression;

public class DoWhileStatement extends Statement {
    private Block block;
    private Expression condition;

    public DoWhileStatement() {
        this.block = new StatementBlock();
    }

    @Override
    public Type getType() {
        return PrimitiveType.Booleano;
    }

    public Block getDoWhileBlock() {
        return this.block;
    }

    public Expression getCondition() {
        return this.condition;
    }

    public void setCondition(Expression condition) {
        this.condition = condition;
    }
}

File: src/main/kotlin/br/com/pnp/model/construct/statement/DoWhileStatement.kt
package br.com.pnp.model.construct.statement

import br.com.pnp.model.construct.Block
import br.com.pnp.model.construct.Variable
import br.com.pnp.model.construct.type.primitive.PrimitiveType
import br.com.pnp.model.expression.Expression

class DoWhileStatement(condition: Expression) : Statement {
    val block: Block = StatementBlock()
    var condition: Expression = Variable.literalBoolean(true)
        set(value) {
            if (!PrimitiveType.boolean.isTypeOf(value.type)) {
                throw IllegalArgumentException(
                        """
                        Type mismatch in ${this.javaClass.simpleName} constructor:
                        Required: ${PrimitiveType.boolean}
                        Found: ${value.type}
                    """.trimIndent())
            }
            field = value
        }

    init {
        this.condition = condition
    }
}
